lib/index.js

```diff
@@ -197,7 +212,15 @@ Metalsmith.prototype.read = unyield(function*(dir){
   dir = dir || this.source();
   var read = this.readFile.bind(this);
   var paths = yield readdir(dir);
-  var files = yield paths.map(read);
+
+  var files = [];
+  var complete = 0;
+  var concurrent = this.concurrent();
+  while (complete < paths.length) {
+    files.push.apply(
+        files, yield paths.slice(complete, complete + concurrent).map(read));
```

Comment:
Can we break this piece apart to be slightly easier to follow? Less one-liner-ish.